bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v2/PWAImpl.java

```diff
@@ -0,0 +1,132 @@
+/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ ~ Copyright 2020 Adobe
+ ~
+ ~ Licensed under the Apache License, Version 2.0 (the "License");
+ ~ you may not use this file except in compliance with the License.
+ ~ You may obtain a copy of the License at
+ ~
+ ~     http://www.apache.org/licenses/LICENSE-2.0
+ ~
+ ~ Unless required by applicable law or agreed to in writing, software
+ ~ distributed under the License is distributed on an "AS IS" BASIS,
+ ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+
+package com.adobe.cq.wcm.core.components.internal.models.v2;
+
+import javax.annotation.Nonnull;
+import javax.annotation.PostConstruct;
+
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ValueMap;
+import org.apache.sling.models.annotations.Model;
+import org.apache.sling.models.annotations.injectorspecific.Self;
+
+import com.adobe.cq.wcm.core.components.models.PWA;
+
+@Model(adaptables = Resource.class,
+    adapters = {PWA.class})
+public class PWAImpl implements PWA {
+
+    static final String CONTENT_PATH = "/content/";
+
+    private boolean isEnabled = false;
+    private String manifestPath = "";
+    private String serviceWorkerPath = "";
+    private String themeColor = "";
+    private String iconPath = "";
+
+    @Self
+    private Resource resource;
+
+    @PostConstruct
+    protected void initModel() {
+        ValueMap valueMap = resource.getValueMap();
+        Boolean isPWAEnabled = valueMap.get(PROP_PWA_PWAENABLED, Boolean.class);
+        this.isEnabled = (isPWAEnabled != null) ? isPWAEnabled : false;
+        if (!this.isEnabled) {
+            return;
+        }
+
+        this.themeColor = colorToHex(valueMap.get(PROP_PWA_THEMECOLOR, ""));
+        this.iconPath = valueMap.get(PROP_PWA_ICON, "");
+
+        String startURL = valueMap.get(PROP_PWA_STARTURL, "");
+        this.manifestPath = replaceSuffix(startURL, MANIFEST_NAME);
+
+        Resource page = resource.getParent();
```

Comment:
We normally use https://helpx.adobe.com/experience-manager/6-2/sites/developing/using/reference-materials/javadoc/com/day/cq/wcm/api/PageManager.html#getContainingPage(org.apache.sling.api.resource.Resource) to obtain the page that contains the component's resource as it avoids hardcoding the node structure depth.